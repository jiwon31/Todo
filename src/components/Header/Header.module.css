.header {
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 1.2rem 0.8rem;
}

.toggle {
  display: flex;
  align-items: center;
  padding: 8px;
  font-size: 1.3rem;
  background-color: transparent;
  transition: all 150ms ease-out;
}

.toggle:hover {
  transform: scale(1.1);
  color: var(--color-accent);
}

.sunny {
  color: var(--color-white);
}

.sunny:hover {
  color: var(--color-accent);
}

.filters {
  display: flex;
  margin: 0;
}

.filter {
  margin: 0.3rem;
  font-size: 1.2rem;
  font-weight: bold;
  text-transform: capitalize;
  color: var(--color-accent);
  opacity: 0.8;
  background-color: transparent;
}

.filter:hover,
.filter.selected {
  opacity: 1;
}

.filter.selected::after {
  content: "";
  display: block;
  margin-top: 0.2rem;
  border-top: 3px solid var(--color-light-grey);
}
