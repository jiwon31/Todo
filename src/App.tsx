import Header from "components/Header/Header";
import TodoList from "components/TodoList/TodoList";
import TodoPresenter from "components/TodoList/todo_presenter";
import { DarkModeProvider } from "context/DarkModeContext";
import { useState } from "react";

type AppProps = {
  present: TodoPresenter;
};
export type Filter = "all" | "active" | "completed";
const filters: Filter[] = ["all", "active", "completed"];

function App({ present }: AppProps) {
  const [filter, setFilter] = useState<Filter>("all");

  return (
    <DarkModeProvider>
      <Header filters={filters} filter={filter} onFilterChange={setFilter} />
      <TodoList present={present} filter={filter} />
    </DarkModeProvider>
  );
}

export default App;
