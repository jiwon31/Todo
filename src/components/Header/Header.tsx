import { Filter } from "app/App";
import { useDarkMode } from "context/DarkModeContext";
import { BsFillMoonFill } from "react-icons/bs";
import { IoSunny } from "react-icons/io5";
import styles from "./Header.module.css";

type HeaderProps = {
  filters: Filter[];
  filter: Filter;
  onFilterChange: (filter: Filter) => void;
};

export default function Header({
  filters,
  filter,
  onFilterChange,
}: HeaderProps) {
  const { darkMode, toggleDarkMode } = useDarkMode();

  return (
    <header className={styles.header}>
      <button className={styles.toggle} onClick={toggleDarkMode}>
        {darkMode ? <IoSunny color="white" /> : <BsFillMoonFill />}
      </button>
      <ul className={styles.filters}>
        {filters.map((value, index) => (
          <li key={index}>
            <button
              className={`${styles.filter} ${
                filter === value && styles.selected
              }`}
              onClick={() => onFilterChange(value)}
            >
              {value}
            </button>
          </li>
        ))}
      </ul>
    </header>
  );
}
