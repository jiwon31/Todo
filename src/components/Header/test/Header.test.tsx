import { Filter } from "app/App";
import Header from "components/Header/Header";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("Header", () => {
  const filters: Filter[] = ["all", "active", "completed"];
  let HeaderComponent: JSX.Element;
  let onFilterChange: () => void;

  beforeEach(() => {
    onFilterChange = jest.fn();
    HeaderComponent = (
      <Header
        filters={filters}
        filter={"all"}
        onFilterChange={onFilterChange}
      />
    );
  });

  it("renders", () => {
    const component = renderer.create(HeaderComponent);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("Filter Click", () => {
    beforeEach(() => {
      render(HeaderComponent);
    });

    it('calls onFilterChange when clicking "all" button', () => {
      const button = screen.getByText("all");
      userEvent.click(button);

      expect(onFilterChange).toHaveBeenCalledTimes(1);
      expect(onFilterChange).toHaveBeenCalledWith(filters[0]);
    });

    it('calls onFilterChange when clicking "active" button', () => {
      const button = screen.getByText("active");
      userEvent.click(button);

      expect(onFilterChange).toHaveBeenCalledTimes(1);
      expect(onFilterChange).toHaveBeenCalledWith(filters[1]);
    });

    it('calls onFilterChange when clicking "completed" button', () => {
      const button = screen.getByText("completed");
      userEvent.click(button);

      expect(onFilterChange).toHaveBeenCalledTimes(1);
      expect(onFilterChange).toHaveBeenCalledWith(filters[2]);
    });
  });
});
