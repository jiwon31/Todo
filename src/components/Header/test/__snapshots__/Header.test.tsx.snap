// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders 1`] = `
<header
  className="header"
>
  <button
    className="toggle"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"
      />
    </svg>
  </button>
  <ul
    className="filters"
  >
    <li>
      <button
        className="filter selected"
        onClick={[Function]}
      >
        all
      </button>
    </li>
    <li>
      <button
        className="filter false"
        onClick={[Function]}
      >
        active
      </button>
    </li>
    <li>
      <button
        className="filter false"
        onClick={[Function]}
      >
        completed
      </button>
    </li>
  </ul>
</header>
`;
